orttraining/orttraining/test/gradient/gradient_checker.cc

```diff
@@ -37,230 +57,164 @@ using training::OpDef;
+                                                       std::vector<std::vector<Y_T>>* y_datas) {
+  op_session.ClearData();
+  for (size_t data_index = 0; data_index < x_datas->size(); ++data_index) {
     std::string name = "input" + std::to_string(data_index);
     const std::vector<X_T>& data = (*x_datas)[data_index];
 
     if (x_infos[data_index].data_type == DataTypeImpl::GetTensorType<int64_t>()) {
       std::vector<int64_t> int64_data(data.size());
       std::transform(data.begin(), data.end(), int64_data.begin(), [](X_T x) { return static_cast<int64_t>(x); });
-      op_session.AddInput<int64_t>(name.c_str(), x_infos[data_index].shape.AsShapeVector(), int64_data);
+      op_session.AddInput<int64_t>(name.c_str(), x_infos[data_index].shape.AsShapeVector(), int64_data, false,
+                                   &x_infos[data_index].dim_params);
     } else if (x_infos[data_index].data_type == DataTypeImpl::GetTensorType<int32_t>()) {
       std::vector<int32_t> int32_data(data.size());
       std::transform(data.begin(), data.end(), int32_data.begin(), [](X_T x) { return static_cast<int32_t>(x); });
-      op_session.AddInput<int32_t>(name.c_str(), x_infos[data_index].shape.AsShapeVector(), int32_data);
+      op_session.AddInput<int32_t>(name.c_str(), x_infos[data_index].shape.AsShapeVector(), int32_data, false,
+                                   &x_infos[data_index].dim_params);
     } else if (x_infos[data_index].data_type == DataTypeImpl::GetTensorType<bool>()) {
       std::unique_ptr<bool[]> p_data(new bool[data.size()]);
       for (size_t i = 0; i < data.size(); ++i) {
         p_data[i] = static_cast<bool>(data[i]);
       }
-      op_session.AddInput<bool>(name.c_str(), x_infos[data_index].shape.AsShapeVector(), p_data.get(), data.size());
+      op_session.AddInput<bool>(name.c_str(), x_infos[data_index].shape.AsShapeVector(), p_data.get(), data.size(),
+                                false, &x_infos[data_index].dim_params);
     } else {
-      op_session.AddInput<X_T>(name.c_str(), x_infos[data_index].shape.AsShapeVector(), data);
+      op_session.AddInput<X_T>(name.c_str(), x_infos[data_index].shape.AsShapeVector(), data, false,
+                               &x_infos[data_index].dim_params);
     }
   }
 
-  for (size_t data_index = 0; data_index < y_infos.size(); data_index++) {
+  for (size_t data_index = 0; data_index < y_infos.size(); ++data_index) {
     std::string name = "output" + std::to_string(data_index);
-    op_session.AddOutput<Y_T>(name.c_str(), y_infos[data_index].shape.AsShapeVector(), (*y_datas)[data_index]);
+    const std::vector<Y_T>& data = (*y_datas)[data_index];
+
+    if (y_infos[data_index].data_type == DataTypeImpl::GetTensorType<int64_t>()) {
+      std::vector<int64_t> int64_data(data.size());
+      std::transform(data.begin(), data.end(), int64_data.begin(), [](Y_T x) { return static_cast<int64_t>(x); });
```

Comment:
not sure I understand this - why do we need this explicit conversion for int64 dtype?

Reply:
Some Ops takes some inputs that are not grad-required in other data type other than float type. The input data generator always generate float vectors, so to set such non-float inputs, we need this conversion. The code here only handle some limit types such as int64, int32, bool, etc, which means that our current GradientChecker can only support Ops that inputs within these types, I am guessing this is enough for now, if someday we want to test a Op that take uint8 as one input type (for example), then we need to add more code to support the new type eventually.

Reply:
I actually move code from L210-L263 from the old file to here as the new AddDatas function. I didn't add such lines by myself (it looks like I did that when comparing to the left from this GitHub review tool...). Eventually if there is any input/output type other than X_T/Y_T, we need such conversion. I am not quite know which Ops will require such int64 conversion, but since it's there, I will keep it. For other places doing the similar thing, if they just don't have such int64 convert previously, when it now call the new AddDatas function, it will just not go into this IF branch, so make no difference.